File: server/datasets/java_source-code/Multiple-C1/organic_8.java
/*
 * Graph BFS - Organic Submission #8
 * Finds the shortest path between two nodes by tracking parents during BFS.
 */

import java.util.*;

class Solution {
    public static List<String> shortestPath(Map<String, List<String>> graph, String start, String target) {
        Map<String, String> parent = new HashMap<>();
        parent.put(start, null);
        Queue<String> queue = new ArrayDeque<>();
        queue.add(start);

        while (!queue.isEmpty()) {
            String node = queue.poll();
            if (node.equals(target)) {
                break;
            }
            List<String> neighbors = graph.getOrDefault(node, Collections.emptyList());
            for (String neighbor : neighbors) {
                if (!parent.containsKey(neighbor)) {
                    parent.put(neighbor, node);
                    queue.add(neighbor);
                }
            }
        }

        if (!parent.containsKey(target)) {
            return null;
        }

        List<String> path = new ArrayList<>();
        String step = target;
        while (step != null) {
            path.add(step);
            step = parent.get(step);
        }
        Collections.reverse(path);
        return path;
    }

    public static void main(String[] args) {
        Map<String, List<String>> graph = new HashMap<>();
        graph.put("A", Arrays.asList("B", "C"));
        graph.put("B", Arrays.asList("A", "D"));
        graph.put("C", Arrays.asList("A", "D"));
        graph.put("D", Arrays.asList("B", "C", "E"));
        graph.put("E", Arrays.asList("D"));

        System.out.println(shortestPath(graph, "A", "E"));
    }
}

File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_c1/organic_8.py
"""
Graph BFS - Organic Submission #8
Finds the shortest path between two nodes by tracking parents during BFS.
"""

from collections import deque


def shortest_path(graph, start, target):
    parent = {start: None}
    queue = deque([start])

    while queue:
        node = queue.popleft()
        if node == target:
            break
        for neighbor in graph.get(node, []):
            if neighbor not in parent:
                parent[neighbor] = node
                queue.append(neighbor)

    if target not in parent:
        return None

    path = []
    step = target
    while step is not None:
        path.append(step)
        step = parent[step]
    path.reverse()
    return path


if __name__ == "__main__":
    graph = {
        "A": ["B", "C"],
        "B": ["A", "D"],
        "C": ["A", "D"],
        "D": ["B", "C", "E"],
        "E": ["D"],
    }
    print(shortest_path(graph, "A", "E"))
